demos/device_defender_for_aws/defender_demo.c

```diff
@@ -0,0 +1,694 @@
+            }
+        }
+
+        if( demoStatus == pdPASS )
+        {
+            /* Note that PublishToTopic already called MQTT_ProcessLoop, therefore
+             * responses may have been received and the publishCallback may have
+             * been called. */
+            for( i = 0; i < DEFENDER_RESPONSE_WAIT_SECONDS; i++ )
+            {
+                /* reportStatus is updated in the publishCallback. */
+                if( reportStatus != ReportStatusNotReceived )
+                {
+                    break;
+                }
+
+                ( void ) ProcessLoop( &mqttContext, 1000 );
+            }
+        }
+
+        if( reportStatus == ReportStatusNotReceived )
+        {
+            LogError( ( "Failed to receive response from AWS IoT Device Defender Service." ) );
+            demoStatus = pdFAIL;
+        }
+
+        /* Unsubscribe and disconnect if MQTT session was established. Per the MQTT
+         * protocol spec, it is okay to send UNSUBSCRIBE even if no corresponding
+         * subscription exists on the broker. Therefore, it is okay to attempt
+         * unsubscribe even if one more subscribe failed earlier. */
+        if( mqttSessionEstablished == 1 )
+        {
+            LogInfo( ( "Unsubscribing from defender topics..." ) );
+            demoStatus = unsubscribeFromDefenderTopics( &mqttContext );
+
+            if( demoStatus == pdFAIL )
+            {
+                LogError( ( "Failed to unsubscribe from defender topics." ) );
+            }
+
+            LogInfo( ( "Closing MQTT session..." ) );
+            ( void ) DisconnectMqttSession( &mqttContext, &networkContext );
+        }
+
+        /* Increment the demo run count. */
+        demoRunCount++;
+
+        if( ( demoStatus == pdPASS ) && ( reportStatus == ReportStatusAccepted ) )
+        {
+            LogInfo( ( "Demo completed successfully." ) );
+
+            /* Reset the flag for demo retry. */
+            retryDemoLoop = pdFALSE;
+        }
+        else
+        {
+            demoStatus = pdFAIL;
+
+            if( demoRunCount < DEFENDER_MAX_DEMO_COUNT )
+            {
+                LogWarn( ( "Demo iteration %lu failed. Retrying...",
+                           ( unsigned long ) demoRunCount ) );
+                retryDemoLoop = pdTRUE;
+
+                /* Clear the flag indicating successful MQTT session establishment
+                 * before attempting a retry. */
+                mqttSessionEstablished = 0;
+
+                LogInfo( ( "A short delay before the next demo iteration." ) );
+                vTaskDelay( DELAY_BETWEEN_DEMO_ITERATIONS_TICKS );
+            }
+            else
+            {
+                LogError( ( "All %lu demo iterations failed.",
+                            ( unsigned long ) DEFENDER_MAX_DEMO_COUNT ) );
+                retryDemoLoop = pdFALSE;
+            }
+        }
+    } while( retryDemoLoop == pdTRUE );
```

Comment:
Nit: Any reason specifically for the `do-while`? Since this is controlled by a counter, I'd think a `for( demoRunCount = 0; demoRunCount < DEFENDER_MAX_DEMO_COUNT; demoRunCount++ )` along with a `break` on success would feel more natural

Reply:
Since there are multiple exit conditions for the loop, its cleaner to log the state this way. If we use condition in for loop, then we have to check why we exited the loop before logging, which means would need to have more state to track.

Reply:
`for` vs `do` is a matter of preference, so it's fine to use one vs the other, but what do you mean by cleaner? The checks would be the same with a `for` loop, but without the `retryDemoLoop` parameter.

Reply:
If we do that, then after the loop we need to check if we exited the loop due to running out of iterations or some other reason, so that we can log that the iterations ran out.

Reply:
I mean, you could change `demoRunCount < DEFENDER_DEMO_MAX_ATTEMPTS` to `demoRunCount < (DEFENDER_DEMO_MAX_ATTEMPTS - 1)`, or loop while <= DEFENDER_DEMO_MAX_ATTEMPTS, or just check outside the loop. I don't think there's anything less clean about it, just different.